onboarding/mlsys/1-backprop.md

```diff
@@ -0,0 +1,414 @@
+background pass, you'll have twice the gradient as before.  If you do it one
+more time, you'll have three times the original gradient.
+
+### Q4:
+
+The additive nature of autodifferentiated gradients is important to
+understanding the features that Determined offers.
+
+Explain in your own words _why_ autodifferentiated gradients are additive.
+
+## Training by Gradient Descent
+
+The basic gradient desent loop algorithm is:
+
+1. Pick an initial model weight.
+2. Select one or more `(x, ylabel)` pairs from the dataset.
+3. Calculate a prediction for the selected inputs with a label (Q2).
+4. Calculate the gradient using the prediction vs label (Q3).
+5. Subtract `gradient * lr` from our model weight ("Minimizing the loss").
+6. Repeat steps 2—5 until you decide to stop.
+
+### Q5
+
+Write a training loop to do gradient descent on our `ypred = mx` model.  Your
+function should take the following inputs:
+
+- initial model weight `m`
+- a learning rate `lr` for calculating the setp
+- a dataset, which is a list of `(x, ylabel)` data pairs
+- a number of iterations to do (one "iteration" is steps 2-5).
+
+Your function should return the final trained weight `m`.
+
+Additionally, in the "select one or more `(x, ylabel)` input pairs" step, for
+now you should just select the next `(x, ylabel)` pair in the dataset, and it
+should wrap back to the previous dataset.  So if your dataset is of length 10,
+then your eleventh iteration should reuse the first `(x, ylabel)` pair of your
+dataset.
+
+Here are some test cases:
+
+```python
+dataset = [(1,1)]
+mfinal = train_loop(m=0, lr=0.001, dataset=dataset, iterations=1)
+assert mfinal == XXX, mfinal
+
+dataset = [(1,1)]
+mfinal = train_loop(m=0, lr=0.001, dataset=dataset, iterations=100)
+assert mfinal == XXX, mfinal
+
+dataset = [(1,1),(2,2),(3,3),(4,4)]
+mfinal = train_loop(m=0, lr=0.001, dataset=dataset, iterations=4)
+assert mfinal == XXX, mfinal
+
+dataset = [(1,1),(2,2),(3,3),(4,4)]
+mfinal = train_loop(m=0, lr=0.001, dataset=dataset, iterations=100)
+assert mfinal == XXX, mfinal
+```
+
+### Q6
+
+What happens if you set the learning rate way too low?
+
+```python
+dataset = [(1,1),(2,2),(3,3),(4,4)]
+mfinal = train_loop(m=0, lr=0.000000001, dataset=dataset, iterations=4)
+print("mfinal", mfinal)
+```
+
+Explain in words what is happening.
+
+### Q7
+
+What happens if you set the learning rate way too high?
+
+```python
+dataset = [(1,1),(2,2),(3,3),(4,4)]
+mfinal = train_loop(m=0, lr=10, dataset=dataset, iterations=4)
+print("mfinal", mfinal)
+```
+
+Explain in words what is happening.
+
+## Batch Size in Training
+
+Your training fuction calculated the gradient for each data point, taking a
+step to minimize the loss for that particular point, then moving on to the next
+point.  This strategy (considering exactly one batch at a time) is called
+"stochastic gradient descent".  It can be an messy process because real gradient
```

Comment:
The "Stochastic" in Stochastic Gradient Descent also implies that the data point (the batch, more generally) is randomly chosen, which isn't the case so far in the example.

Reply:
good to know, I'll incorporate that somehow.

Reply:
avoided SGD in the rewrite.